Improve Nav block loading and placeholder states

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Currently the loading and placeholder states for the Nav block are quite difficult to reason about. 

- There's a lot of component nesting and states within states. An example is the `PlacholderPreview` which is rendered at both the root block level _and_ within the `PlaceholderComponent` itself. 
- The logic determining when the placeholder is shown requires setting state when this can simply be derived from existing state.
- The loading states are too complex and obscured.
- The loading state visuals are difficult to distinguish.
- Little or no non-visual communication of loading progress.

This PR seeks to solve the above issues by:

- deriving the placeholder visibility from existing state rather than requiring `setState()`s.
- add a clear "loading" spinner to the loading state as discussed with @jasmussen.
- consolidate all loading state logic into the block _root_ - leaving only "locally relevant" loading logic (i.e. resolution status of which optoins are available within the placeholder actions) within the Placeholder component.
- avoid potential for overlapping rendering by returning distinct components early for both placeholder and loading states
- adds `aria-live` announcement of block loading progress.
- avoid rendering both placeholder preview and block appender when the selected menu is empty

🙏 **Please note**:  this PR does not attempt to fix the "freezing" of the UI that occurs when you create a new Menu or select an existing one. That issue is addressed in https://github.com/WordPress/gutenberg/pull/38858. The aim of this PR is simply to iterate towards an improved loading state for the block.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

I recommend testing the scenario on `trunk` first to get a sense of just how suboptimal the current loading experience is.

Now on this PR's branch:

1. Create new Post.
2. Set Devtools to `Slow 3G` and add a Nav block.
3. Notice the loading spinner.
4. Choose an existing Menu with items.
5. Save and reload page.
6. See loading spinner whilst items are being resolved.
7. Add a new Nav block and choose create _empty_.
8. Save and reload page.
9. See loading state whilst items are resolved and then see the placeholder (this shows because there are no items in the menu/block).

## Screenshots <!-- if applicable -->

#### Before

No clear loading state for initial setup or subsequent loading of existing blocks:

https://user-images.githubusercontent.com/444434/154667604-529a9879-2394-42b1-a949-faa8361f6adf.mov


#### After

Clear loading state for initial setup _and_ subsequent loading of existing blocks:


https://user-images.githubusercontent.com/444434/155123560-cd7a43b4-e1f2-4d2f-893e-f8ff31d1ae41.mov


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Improve all loading statess

jasmussen commented:
Thank you for taking a stab at this. Even as is, this is a big improvement. Here are a few gifs. Insert fresh:
![insert fresh](https://user-images.githubusercontent.com/1204802/154662317-211d0f20-3131-457b-bf89-c2c253062d95.gif)

_Nice_. 👌  

(I hope to look separately at making sure the _height_ of this placeholder state is the same as a menu set in that theme will be, right now it's locked to 48px, which can cause some shifting.)

Selecting a menu:

![selecting a menu](https://user-images.githubusercontent.com/1204802/154662443-ce184597-38e0-4745-a80f-182e658df982.gif)

Blink and you'll miss it. It does look like the spinner has a black border around it, that'd be nice to do without. Possible?

Reloading (or loading fresh) a page that has a specific navigation already:

![loading a menu](https://user-images.githubusercontent.com/1204802/154662641-69936364-0fe5-4c22-89b7-a0349c90f943.gif)

 This is probably the one that could use the most love. Doesn't all have to be in a single PR, I know how deep the rabbit hole goes as far as the code — and I'm happy to help here. Seems like the dream scenario could be:

- Don't show the dashed line placeholder at all if a menu is to be loaded. That might mean deferring the existance of that placeholder until _after_ the initial loading, maybe?
- Solo spinner and dashed-line-placeholder both always have the same height as an eventual menu item would have, to avoid layout shifts. I can probably address this one separately with some tricks.

jasmussen commented:
Tons of work in a short time, wild 👌

Took it for another spin. Inserting:

![inserting](https://user-images.githubusercontent.com/1204802/154678196-e1ae84f7-0759-46b5-bf88-9a2211c587e1.gif)

Cool. I wonder if we need to see the spinner there at the end or not, seems like we could do without, but that's not a strong opinion.

Selecting a menu from the placeholder:

![selecting](https://user-images.githubusercontent.com/1204802/154678275-d6f32d18-3caf-48dc-85a5-4876d7ef195c.gif)

Nice. There's still a black border around the blink and you'll miss it spinner appearing. As noted, I'd like to try separately to fix the height shifting from placeholder to spinner to menu, so it's one height for all three.

Reloading:

![reloading](https://user-images.githubusercontent.com/1204802/154678396-c42e8865-c285-43fd-9b09-fecfa4bde75d.gif)

Probably still the most difficult flow to get right, and it's looking very close! Basically it looks like there are _two_ spinners now, first an initial basic one, then one inside a bordered box, and then we get the menu. If we could reduce that to just the first spinner, seems like it'd be a shoo-in!

getdave commented:
Thanks for super fast reviews 🙇 

> There's still a black border around the blink and you'll miss it spinner appearing.

I cannot see the black border. Could you try clean `npm i` and `npm run dev`?

> Probably still the most difficult flow to get right, and it's looking very close! Basically it looks like there are two spinners now, first an initial basic one, then one inside a bordered box, and then we get the menu. If we could reduce that to just the first spinner, seems like it'd be a shoo-in!

I believe I have fixed this with latest commit.

getdave commented on packages/block-library/src/navigation/edit/inner-blocks.js:
```diff
@@ -96,7 +96,7 @@ export default function NavigationInnerBlocks( {
 		( isImmediateParentOfSelectedBlock && ! selectedBlockHasDescendants );
 
-	const placeholder = useMemo( () => <PlaceholderPreview />, [] );
+	const placeholder = PlaceholderPreview;
```
This appeared to be causing the component to be rendered when it didn't need to be.

I'm not sure this component is rendered often enough to warrant a memo here but I could be wrong.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -658,29 +693,6 @@ function Navigation( {
 					</InspectorControls>
 				) }
 				<nav { ...blockProps }>
-					{ isPlaceholderShown && (
```
These complex loading states have been extracted to separate conditionals each of which returns early with a dedicated render.

getdave commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -39,79 +46,91 @@ export default function NavigationPlaceholder( {
+
+						<hr />
+
+						{ isResolvingActions && <Spinner /> }
```
Depending on UX this may not be required. Depends if we're happy to show an empty placeholder whilst we wait for the action buttons to resolve (or not).

Remember on slow connections you might see an empty placeholder for quite some time...

jasmussen replied:
Transversal idea: could we build in a behavior in the spinner (default or optional) where the spinner only appears after, say, 2 seconds of inactivity? That would make it "optimistic", only appear when a transaction is actually slow. 🤔

draganescu replied:
Love @jasmussen 's idea but ideally in another PR since spinner is from the components package.

talldan replied:
Definitely agree with this. There was a related issue here - https://github.com/WordPress/gutenberg/issues/38970.

It'd be good to apply the optimism to Page List's loading spinner in the separate PR as well. I think a lot of the time users are seeing that and mistaking it for the Nav Block's loading state.

getdave replied:
So one for another PR then.

getdave commented on packages/block-library/src/navigation/edit/placeholder/placeholder-preview.js:
```diff
@@ -9,12 +9,11 @@ import classnames from 'classnames';
 import { Icon, navigation } from '@wordpress/icons';
 import { __ } from '@wordpress/i18n';
 
-const PlaceholderPreview = ( { isLoading } ) => {
```
Loading variant seemed confusing. Not sure why it was necessary.

talldan replied:
`PlaceholderPreview` previously displayed the gray empty state blobs, and `isLoading` was the pulse animation variant of that.

Feels like this entire component is redundant now. It does the same thing as the Placeholder, but without any buttons. Maybe it can be deleted, but I haven't checked to see where it's used.

getdave replied:
> Not sure why it was necessary.

I should have qualified that statement. I understood why it existed but not it's no longer required due to refactoring.

> Maybe it can be deleted, but I haven't checked to see where it's used.

It _was_ also used as the `placeholder` prop arg for the `<InnerBlocks>` component which would show when the inner blocks were _empty_ but present and _not selected_.

I'll check whether it can be ditched. It would certain make things less confusing 😄

getdave replied:
I checked and it's used in the following scenarios:

1. The block is _unconfigured_ and not currently selected.
2. The block is _configured_ (i.e. has `<InnerBlocks>`) but _empty_ and _not currently selected_.

I think I would prefer to create a follow up whereby we refactor this component away. Trying to do it in this PR is going to create a fair amount of noise which I think will make it even harder to have confidence in the PR.

How do you feel about that? Happy to raise the Issue.

jasmussen commented:
Thanks for your persistence here! I've pulled the latest, I've npm installed, cleared cache, using chrome, but I'm still seeing a few black borders in the reload and select cases:
![reload](https://user-images.githubusercontent.com/1204802/154695486-fdad0efd-d840-4a1f-9d47-8243120b0a84.gif)

![selecting](https://user-images.githubusercontent.com/1204802/154695497-5b0eb688-7a5b-469f-8e5f-ed46e0af9fda.gif)

getdave commented:
https://user-images.githubusercontent.com/444434/154722900-6e10d82d-8b66-4121-9cea-193e46d1a718.mov

For reference here's what I see. I have blue outlines which makes me think this is a Theme specific thing. Perhaps something is modifying the outline color?

Note that the `<Spinner />` is wrapped in a `<nav>` which has `blockProps` spread onto it's props. This ensures that layout is consistent. However I imagine that the Nav has a "selected" state and perhaps that is what we're seeing?

tellthemachines commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -39,79 +46,91 @@ export default function NavigationPlaceholder( {
+
+						{ isResolvingActions && <Spinner /> }
+
+						{ showSelectMenus ? (
```
Now that we've removed the "Add all pages" option, does it even make sense to show the placeholder if there are no existing menus to select? It feels a bit weird to have a placeholder with a single option inside it:

<img width="871" alt="Screen Shot 2022-02-21 at 5 27 34 pm" src="https://user-images.githubusercontent.com/8096000/154900729-aa4d32a3-74cd-4792-bc92-8f43adaa13ea.png">

jasmussen replied:
This is an excellent and pertinent question touching on a larger issue that'd be good to explore. Overall, it would be very nice if we could rid of the placeholder in most cases, such as if you already have a single menu defined, then activate a new theme, just load that first menu automatically, since it's so easy to swap it out from the toolbar anyway.

The other is the behavior of when _zero_ menus are defined. If I understand this correctly, the "Page List" block is currently output automatically on frontends in empty menus, is that right? While useful in principle it does make for a bit of a disconnect between frontend and editor, that could result in confusion: "I'm seeing all my pages here, but not there". In that light, I wonder if we should just preinsert the Page List block in navigation, if no saved menus are created, so the two are in sync? I'm assuming no, based on past discussions on navigation persistence, but I'm suggesting the concept in case it can inspire other ideas. We still don't really have a good way to create patterns for the navigation either.

The new empty state is inspired by Site Logo and Featured Image, and features this dashed line:

<img width="683" alt="Screenshot 2022-02-21 at 10 26 48" src="https://user-images.githubusercontent.com/1204802/154926491-8042a370-eb0b-4e20-88c7-16d08062f24a.png">

Outside of somehow pre-populating an empty block, it seems prudent to at least keep that state around to indicate: there's an empty navigation block here which you can click to configure. If we can have that state, and go directly to the nav inserter on select, in principle that sounds good. But wouldn't that require us to defer the saving to a CPT? Otherwise simply clicking the dashed line nav block would presumable create an empty menu which was then loaded by default? 🤔

jasmussen commented:
> Note that the <Spinner /> is wrapped in a <nav> which has blockProps spread onto it's props. This ensures that layout is consistent. However I imagine that the Nav has a "selected" state and perhaps that is what we're seeing?

I'm still seeing the black border, which suggests I'm seeing an unstyled `Placeholder` component with the spinner inside it. I'm also using TT2.

Can you try the use case where you reload a page with a selected menu?

Not sure what I'm doing wrong here 🤔

alexstine commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -39,79 +46,91 @@ export default function NavigationPlaceholder( {
 
+	useEffect( () => {
+		if ( isResolvingMenus ) {
+			speak(
```
@getdave Does this only fire on selectBlock()? If not, users with more than 1 Navigation Block could be in for a not so great experience with multiple messages being read.

getdave replied:
Great catch. I've added this. One to remember for the future. Thank you 🙇‍♂️

alexstine reviewed (commented):
@getdave Looking good for accessibility. Just 1 question below.

jasmussen commented:
Alright, with the help of Dave in a chat debugging session, it became clear that the black border I was seeing was from the Page List block. Which of course I should've probably clarified in my testing steps — my apologies, a beverage is owed by yours truly.

All that is to say: I'm now seeing the same:
![loading](https://user-images.githubusercontent.com/1204802/155118511-0c51bba3-8213-4815-b7b3-54b25052ee00.gif)

And this is a vast improvement, and not one that should be held up by potential separate Page List block improvements. Thanks Dave, this is good for me.

_I do still separately want to look at adding a min-height based on menu item text, so we can avoid layout shifts (and find a better alternative to #38439), but that's also separate._

jasmussen reviewed (approved):
Per the realization in https://github.com/WordPress/gutenberg/pull/38907#issuecomment-1047674832, I gave this another test, and it's solid. 

GIF showing how to insert fresh:
![insert fresh](https://user-images.githubusercontent.com/1204802/155122377-ecb008a0-873f-4f81-b431-3874fb58bcf9.gif)

GIF showing how to select between menus (note, the black border flashing is a Page List):
![select](https://user-images.githubusercontent.com/1204802/155122427-33dc7cbd-d18e-424d-a471-69ab0e2af084.gif)

GIF showing reloading a page with a navigation:
![reload](https://user-images.githubusercontent.com/1204802/155122455-a3c2ab66-da79-4ef4-b938-99c09777ce89.gif)

GIF showing "Start empty":
![start empty](https://user-images.githubusercontent.com/1204802/155122517-3037ef8b-0191-4c66-8830-579955a8baa2.gif)

This is a vast improvement, and should fix the glitching in #38970. 

I'll follow up and create tickets for:

* Page List loading state followup
* Avoiding layout shifts between placeholder and final menu
* Use a white plus button in an empty nav block

jasmussen commented:
Created a trio of tickets:

* #38980
* #38981
* #38982

getdave commented:
Thank you for the review and all your work @jasmussen ✨ 

Now seeking a technical review 👀

getdave commented:
Rebasin'

draganescu commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -186,7 +186,7 @@ function Navigation( {
 					( block ) => block.name === 'core/navigation-submenu'
 				),
-				hasUncontrolledInnerBlocks: _hasUncontrolledInnerBlocks,
+				hasUncontrolledInnerBlocks: !! _hasUncontrolledInnerBlocks,
```
Why the `!!`? Could we fix `_hasUncontrolledInnerBlocks` to be definitely bool?

getdave replied:
These values are derived from values which a coerced to Booleans but are in fact empty arrays or numbers.

I just refer the clarity of `hasX` definitely being a Boolean rather than some other value.

> Could we fix _hasUncontrolledInnerBlocks to be definitely bool?

We'd just end up with

```js
const _hasUncontrolledInnerBlocks =
				!! _uncontrolledInnerBlocks?.length;
```

talldan replied:
This does look like an oversight (probably by me!)—good idea to add the `!!`. 👍 

I agree with @draganescu that it would be good to change the `_has` variable to boolean like the name implies.

getdave replied:
✅ Fixed in https://github.com/WordPress/gutenberg/pull/38907/commits/092947d23a0bc2d8ac260b5d808a84b24af35403

draganescu reviewed (approved):
Outside of minor tweaks which could be addressed separately I think this looks good.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -476,6 +477,36 @@ function Navigation( {
+	if ( isLoading ) {
+		return (
+			<nav { ...blockProps }>
+				<Spinner className="wp-block-navigation__loading-indicator" />
```
`className` added to distinguish from other `<Spinner>` instances.

getdave commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -43,9 +43,10 @@ export default function NavigationMenuSelector( {
-		( menu ) => ! currentMenuId || menu.id !== currentMenuId
-	);
+	const navigationMenusOmitCurrent =
+		navigationMenus?.filter(
```
Same as `trunk` except I added a guard against `navigationMenus` being "empty" or `null`.

getdave committed:
Translate spoken strings

getdave commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -1,94 +1,117 @@
+		}
+	}, [ isResolvingMenus, isSelected ] );
+
+	const isResolvingActions =
```
Only "resolving" in the UI if _both_ of these are resolving. If one is completed then we can show the user something already.

adamziel replied:
This doesn't hide anything, though, right? It only shows or hides the spinner.

getdave replied:
Basically. It's a low fidelity approach to avoid there being nothing showing on slower connections whilst the resolution states resolve.

If you recall we're going to try and refactor the logic out of `NavigationMenuSelector` into a hook. But in a follow up PR.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -301,6 +296,18 @@ function Navigation( {
 
+	// "placeholder" shown if:
+	// - we don't have uncontrolled blocks.
+	// - (legacy) we have a Navigation Area without a ref attribute pointing to a Navigation Post.
```
That's a good reminder to remove nav area code. 😄

adamziel replied:
+1, I think the grace period for the deprecation is now over

getdave replied:
Ok so I can just remove this? Don't need to do anything else?

getdave replied:
On second thoughts, I'd rather not mix up the two goals. I've created a follow up https://github.com/WordPress/gutenberg/issues/39128

talldan replied:
Good to do it separately 👍

talldan commented on packages/block-library/src/navigation/edit/inner-blocks.js:
```diff
@@ -98,6 +97,15 @@ export default function NavigationInnerBlocks( {
 
+	const hasMenuItems = !! blocks?.length;
+
+	// If here is a `ref` attribute pointing to a `wp_navigation` but
```
```suggestion
	// If there is a `ref` attribute pointing to a `wp_navigation` but
```

getdave replied:
✅

talldan commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -43,9 +43,10 @@ export default function NavigationMenuSelector( {
-	const navigationMenusOmitCurrent = navigationMenus.filter(
-		( menu ) => ! currentMenuId || menu.id !== currentMenuId
-	);
+	const navigationMenusOmitCurrent =
+		navigationMenus?.filter(
+			( menu ) => ! currentMenuId || menu.id !== currentMenuId
+		) || [];
```
If the current menu id isn't set it looks like there's an opportunity to avoid the entire filter operation.

Here's a rough sketch to show what I mean:
```suggestion
let navigationMenusOmitCurrent = navigationMenus ?? [];

if ( currentMenuId ) {
   navigationMenusOmitCurrent =
	    navigationMenusOmitCurrent.filter( ( menu ) => menu.id !== currentMenuId );
}
```

Then the optional chaining can also be dropped because the menus have a default using `??`.

----

The other UI option is that we continue to show the currently selected menu in the dropdown, but indicate that it's selected by showing a check next to it. That's how it used to work via the `value` prop of `MenuItemsChoice`:
https://github.com/WordPress/gutenberg/blob/9a58337f7e04e95fe54f636fbddf62821d4d79ce/packages/block-library/src/navigation/edit/navigation-menu-selector.js#L18-L31

IMO it's good to also show what the currently selected menu is somewhere in the UI, as it isn't shown anywhere, but granted this is subjective. The menu items would have to use the role `menuitemradio` when there's a selected menu if that change were made. I think that's semantically correct though, because they are radio buttons.

tellthemachines replied:
I'm inclined to show the selected menu with a check next to it; it makes it clear what is currently selected (someone might go looking for their "header menu" not realising it's already selected) and also because removing the selected item from the dropdown causes the whole component to re-render and focus (which should go back to the Select Menu button after close) gets lost in the process.

getdave replied:
I've refactored to use the selected menu with a check. The semantics of a radio suit the scenario much better and avoid the re-rendered we mentioned. This is much better thank you 🙇‍♂️ 🙇‍♂️ 🙇‍♂️

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -1,94 +1,117 @@
+
+		if ( isResolvingMenus ) {
+			speak(
+				__( 'Loading Navigation block setup placeholder options.' ),
```
We can probably drop one of either 'setup' or 'placeholder' from this text and the text just below this. 

I think probably 'placeholder' can go because that's what we call it internally, but it's not something every user will understand (and it's often used in a different way).

```suggestion
				__( 'Loading Navigation block setup options.' ),
```

What do you think?

getdave replied:
Yeh `placeholder` makes no sense to users. It's a developer centric term 👍

getdave replied:
✅

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -1,94 +1,117 @@
+		if ( hasResolvedMenus ) {
+			speak(
+				__( 'Navigation block setup placeholder options ready.' ),
+				'polite'
```
I think 'polite' is the default, so it doesn't need to be specified.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -1,94 +1,117 @@
-							</div>
+			<Placeholder className="wp-block-navigation-placeholder">
+				{
+					// The <PlaceholderPreview> component is displayed conditionally via CSS depending on
+					// whether the block is selected or not. This is achieved via CSS to avoid
+					// component re-renders
```
I couldn't see how that might be achieved, there aren't any conditional classnames added to `PlaceholderPreview`.

getdave replied:
I believe it's via the cascade. I can try and dig it out.

getdave replied:
Here we go. It's shown/hidden based on whether block is selected.

https://user-images.githubusercontent.com/444434/156017233-dd5c620a-bf04-477b-8e29-d024d4abd695.mov

https://github.com/WordPress/gutenberg/blob/3efbac2056858211aeda54e2ba8751817d9aaee7/packages/block-library/src/navigation/editor.scss#L272-L274

adamziel replied:
Ideally the comment would answer that question. Can we update it to say this hinges on the parent element having an `.is-selected` class?

talldan commented on packages/block-library/src/navigation/edit/placeholder/placeholder-preview.js:
```diff
@@ -9,12 +9,12 @@ import classnames from 'classnames';
-				'wp-block-navigation-placeholder__preview',
-				{ 'is-loading': isLoading }
+				'wp-block-navigation-placeholder__preview'
 			) }
```
Looks like the call to the `classnames` function can be removed now,

getdave replied:
✅

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -1,94 +1,117 @@
-								</Button>
-							) }
-						</div>
+						<hr />
```
Not introduced by this PR, but looks like a quick fix.

Should the `<hr />` be inside the `canUserCreateNavigationMenu`  conditional below?

Otherwise it's shown when there's potentially no button after the `NavigationMenuSelector`.

getdave replied:
@adamziel and I discussed this. Basically the conditionals around `NavigationMenuSelector` need to be extracted and made into a hook as they are needed in several places. That's going to be addressed in another PR. For now we're currently fixing this by using CSS to hide any adjacent `<hr />`s.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -301,6 +296,18 @@ function Navigation( {
+	// - we don't have uncontrolled blocks.
+	// - (legacy) we have a Navigation Area without a ref attribute pointing to a Navigation Post.
+	// - we don't have a ref attribute pointing to a Navigation Post.
+	const isPlaceholderShown =
+		!! ( ! hasUncontrolledInnerBlocks || isWithinUnassignedArea ) && ! ref;
```
Changing this makes me really nervous, as `setIsPlaceholderShown` was definitely a fragile bit of code in the past.

I think we just need to test this very thoroughly. The nav block has pretty good e2e tests now, so that brings a bit more confidence 👍

adamziel replied:
That's a lot of operations and I'm a bit lost here. Why do we need all the boolean coercions? Shouldn't the one here apply to the entire expression and not just to the first parenthesis?

getdave replied:
So let's ignore the `isWithinUnassignedArea` as we can deprecate it.

The statement is

- if there are no uncontrolled inner blocks (i.e. a pattern or a legacy nav block). _**OR...**_
- if there is no `ref` pointing to a `wp_navigation` post 

...then it's safe to show the placeholder.

Previously we were setting state to determine whether the placeholder was shown. AFAIK this was unnecessary overhead when the state could be derived.

All the e2e tests continue to pass and @jasmussen has been over this a few times.

This piece also makes me the most nervous @talldan. That's why I felt more eyes on it the better. 

Let me simplify this and see if it improves our confidence.

getdave replied:
I've removed the coercion and reversed the state ordering. Once we drop `isWithinUnassignedArea` (in a follow up) the statement will read

```js
const isPlaceholderShown = ! ref && ! hasUncontrolledInnerBlocks;
```

...which is easier to reason about.

Here's the code on trunk for comparison

https://github.com/WordPress/gutenberg/blob/3efbac2056858211aeda54e2ba8751817d9aaee7/packages/block-library/src/navigation/edit/index.js#L208-L210

https://github.com/WordPress/gutenberg/blob/3efbac2056858211aeda54e2ba8751817d9aaee7/packages/block-library/src/navigation/edit/index.js#L335-L337

getdave replied:
Note I reviewed the Template Part block and they also _avoid_ using `setState` to control the placeholder, preferring instead to use _derived state_ as I have here.

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -264,7 +264,131 @@ describe( 'Navigation', () => {
+		} );
+	} );
+
+	describe( 'placeholder states', () => {
```
I don't know that we need two `describe`s for the placeholder, feels like it could just be one.

getdave replied:
Yeh I guess I'm refactoring tests. Let's group in one as you suggest.

getdave replied:
Merged into one with sub `describe` blocks for clarity.

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -264,7 +264,131 @@ describe( 'Navigation', () => {
+				'.editor-post-text-editor'
+			);
+			await codeEditorInput.click();
+			const markup = '<!-- wp:navigation /-->';
```
Do we need to use the code editor for this? `<!-- wp:navigation /-->` is the nav block's markup when it's normally.

getdave replied:
Yes it's correct that this is the block markup when inserting a Nav block for the first time. I was trying to test the loading state of the block when it isn't selected. If you insert the block it will be selected which is different.

I could have different tests. One which inserts a block and reloads the page and asserts on the loading results. And another which checks on the behaviour when the block is initially inserted. 

What do you think?

It's all about ensuring we don't see loading indicators in either selected or unselected states.

talldan replied:
Good point about it not being selected. Maybe add a comment to the test to mention that's the reason behind using the code editor.

getdave replied:
✅

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -264,7 +264,131 @@ describe( 'Navigation', () => {
+
+			const testNavId = 1;
+
+			await setUpResponseMocking( [
```
Another option instead of mocking is network throttling:
https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pageemulatenetworkconditionsnetworkconditions

Response mocking probably still offers more control though, so I don't really have a preference. It would be good to add some comments about this line to explain the purpose.

getdave replied:
I would definitely promote controlling the test flow explicitly via mocking. I will explain this thoroughly in comments now.

getdave replied:
I've added a load more comments to explain my reasoning.

talldan reviewed (commented):
Thanks for undertaking this change 👍 

There were a few comments, but they were all quite small and I think fairly easy to address.

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -476,6 +477,36 @@ function Navigation( {
 		{ open: overlayMenuPreview }
 	);
 
+	if ( isLoading ) {
+		return (
+			<nav { ...blockProps }>
+				<Spinner className="wp-block-navigation__loading-indicator" />
+			</nav>
+		);
+	}
```
I chatted just now to @tellthemachines about https://github.com/WordPress/gutenberg/issues/38169, and I'd be concerned this early return won't help matters because it'll cause `NavigationMenuSelector` to be unmounted when switching menus.

adamziel replied:
Perhaps this would be an easy way out:

```js
let Loader;
if ( isLoading ) {
	Loader = (
```

and then below render it as `{ Loader }` ?

getdave replied:
Oh I see. Well we have a problem then because from a design perspective @jasmussen was hoping that the block would visually just show the spinner when "loading" and nothing else.

So perhaps we show the spinner but keep the block toolbar mounted but visually hidden?

tellthemachines replied:
Showing only the spinner might make sense when a block is first added, or page first loads, but it feels a bit weird when the only thing we're doing is selecting a new menu to render in that place. I wouldn't even hide the toolbar in that case, I'd expect it to stay in place while the nav contents are switched up. After all, the only change is in the content, the block instance stays the same.

jasmussen replied:
Sorry for missing the ping here! One utopian version is we never show a spinner because things load so fast there's no need to indicate "processing". It's the same motivation that prompted the idea to defer the spinner to only appear if 2 seconds have passed.

Forgive me if I'm missing some technical aspects here, but if we are able to swap out the contents of a menu without showing a spinner (for instance by keeping the old menu present until the new one is able to take its place), that sounds great, especially as it would avoid layout shifts. Did I get that right?

getdave replied:
Ok I've changed the implementation so that the loading state isn't an early return. This means the spinner is rendered alongside the toolbars if they are available.

Note that this doesn't fix https://github.com/WordPress/gutenberg/issues/38169 because that appears to be a more complex problem but at least it won't make things worse. Thanks @tellthemachines for catching this one.

We'll need to add some e2e test coverage in any fix for https://github.com/WordPress/gutenberg/issues/38169 to ensure we don't have regressions.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -301,6 +296,18 @@ function Navigation( {
+	// "loading" state:
+	// - there is a ref attribute pointing to a Navigation Post
+	// - the Navigation Post hasn't resolved yet.
+	const isLoading = !! ( ref && ! isEntityAvailable );
```
What if we omit the `!!` and just let the `if ( isLoading ) {` do the coercion implicitly?

getdave replied:
~I think in this case we don't need explicit coercion~ Scratch that @adamziel. My preference is for `hasX` variables to always been explicitly a `Boolean`. I prefer not to rely on implicit coercion because it requires the developer reading the code to have to think about it. 

In fact given the choice I would do this to be ultra clear:

```js
const isLoading = Boolean( ref && ! isEntityAvailable );
```

Is there any harm in the `!!` when it's been established as a proxy for `Boolean`? Perhaps there's a problem I missed?

adamziel replied:
My point was that there is no escaping the coercion. Whether in the variable declaration using `!!` or in an `if` condition, it is happening. If you want to the variable to be a boolean from the start, then how about something like `typeof ref === "string" && ! isEntityAvailable`  or `ref !== null && ! isEntityAvailable` or even `!! ref && ! isEntityAvailable` without the wrapping parenthesis?

adamziel replied:
>  I prefer not to rely on implicit coercion because it requires the developer reading the code to have to think about it.

Does moving the coercion from the `if` to `const` removes the need to think about it?

adamziel commented:
I see a brief flash of the placeholder style when I'm switching menus, perhaps related to https://github.com/WordPress/gutenberg/pull/38907/files#r813555589

adamziel commented:
I noticed the "Select menu" label sometimes disappear from the toolbar when I select a different menu. It seems to be related to requesting the entity record details.

![CleanShot 2022-02-25 at 18 47 27](https://user-images.githubusercontent.com/205419/155762785-94803151-f56c-4679-a84e-9ded9efb5e6f.gif)

Perhaps it's the same thing @talldan [brought up](https://github.com/WordPress/gutenberg/pull/38907/files#r813555589)

getdave committed:
Coerce to Boolean at source

Addresses https://github.com/WordPress/gutenberg/pull/38907#discussion_r813492976

getdave committed:
Correct typo

getdave committed:
Refactor to use radio control component for block based menu selector

getdave committed:
Group placeholder tests

getdave committed:
Remove unnecessary usage of classnames util

getdave committed:
Fix inner blocks placeholder to be visible by ARIA

getdave committed:
Only set aria-hidden where necessary

getdave committed:
Remove developer centric term from placeholder announcements

Addresses https://github.com/WordPress/gutenberg/pull/38907#discussion_r813504821

getdave committed:
Remove default polite arg to speak function

Addresses https://github.com/WordPress/gutenberg/pull/38907#discussion_r813505319

getdave committed:
Reverse conditional to aid in clarity

getdave committed:
Add clarity to test purpose and method via comments

Addresses https://github.com/WordPress/gutenberg/pull/38907#discussion_r813532854

getdave committed:
Improve test comprehension via comments

Addresses https://github.com/WordPress/gutenberg/pull/38907#discussion_r813535940

getdave commented:
> I noticed the "Select menu" label sometimes disappear from the toolbar when I select a different menu. It seems to be related to requesting the entity record details.
> 
> ![CleanShot 2022-02-25 at 18 47 27](https://user-images.githubusercontent.com/205419/155762785-94803151-f56c-4679-a84e-9ded9efb5e6f.gif)
> 
> Perhaps it's the same thing @talldan [brought up](https://github.com/WordPress/gutenberg/pull/38907/files#r813555589)

@adamziel I think it's because the `ref` is reset and temporarily not resolved. Therefore this conditional causes the selector to unmount.

https://github.com/WordPress/gutenberg/blob/3efbac2056858211aeda54e2ba8751817d9aaee7/packages/block-library/src/navigation/edit/index.js#L483-L488

Any suggestions 🤔 ?

talldan commented:
Yeah, I think there are a few causes, mostly conditional rendering. But then there's also this that I spotted - https://github.com/WordPress/gutenberg/issues/39044.

I don't think we need to fix everything here though, if it's the same in `trunk` it might be an idea to focus on fixing it in a separate PR.

getdave committed:
Qualify test comment to explain purpose

getdave committed:
Update snapshots

getdave committed:
Remove early return for loading state

getdave committed:
Reduce verbosity of vars and colocate

getdave commented:
As @talldan suggests let's not try and fix the focus loss in this PR 👍 . We'll just make sure we're not making it worse!

scruffian commented:
The spinner is a bit hard to see if you're in a dark theme:
![Mar-01-2022 10-59-15](https://user-images.githubusercontent.com/275961/156157206-a45d969f-807d-414f-a8ad-dfe35c35fb6b.gif)

getdave commented:
@scruffian Seems to be ok except when it's selected. Not sure how to avoid that as we need the selected visual cue for a11y reasons
![Screen Capture on 2022-03-01 at 11-51-06](https://user-images.githubusercontent.com/444434/156164428-7bc8bd08-a0ac-43f7-9f5c-98e7b2f32837.gif)

scruffian commented:
Could we add a bit of padding around the spinner to make a clearer visual distinction between the spinner and the selected state?

adamziel commented:
Is that the one from `@wordpress/components`? It might be a good opportunity to make it more dark more-friendly cc @jasmussen.

jasmussen commented:
There's a visual present in #36725 for the spinner could work in dark contexts. To put it shortly: white instead of spot-color blue, but keep the track the same color. Could potentially be a small PR that just leverages the existing `is-dark-theme` class that gets applied in canvas? I.e. pseudo code:

```
.components-spinner-indicator {
	color: blue;

	.is-dark-theme & {
		color: white;
	}
}
```

getdave commented:
> Could we add a bit of padding around the spinner to make a clearer visual distinction between the spinner and the selected state?

How about this?

![Screen Capture on 2022-03-01 at 13-42-55](https://user-images.githubusercontent.com/444434/156179869-aec0c609-2d71-46f9-be96-cb4ac0797af7.gif)

getdave committed:
Add spacing around spinner when block is selected

getdave commented:
> Is that the one from @wordpress/components? It might be a good opportunity to make it more dark more-friendly cc @jasmussen.

One for _another_ follow up I think.

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -648,56 +673,40 @@ function Navigation( {
+					</TagName>
+				) }
+
+				{ ! isLoading && ! isPlaceholder && (
```
Nit: isPlaceholder is certain to be false at this point.

getdave replied:
✅

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -648,56 +673,40 @@ function Navigation( {
+
+				{ ! isLoading && ! isPlaceholder && (
+					<TagName { ...blockProps }>
+						{ ! isPlaceholder && (
```
Nit: isPlaceholder is certain to be false at this point. If we're going to remove the `if() { return; }` above, it might still make sense to keep this here, though.

getdave replied:
✅ Removed. If we need it back the tests will provide coverage.

adamziel commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -98,26 +93,32 @@ export default function NavigationMenuSelector( {
-									</MenuItem>
-								);
-							} ) }
+							<MenuItemsChoice
```
this is great

adamziel commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -98,26 +93,32 @@ export default function NavigationMenuSelector( {
+										)
+									);
+								} }
+								choices={ navigationMenus.map(
```
nit: useMemo would be nice for that to avoid passing in a different object each time

getdave replied:
I don't have any evidence to support this but given that `MenuItemsChoice` could potentially render a number of items it could make sense to optimise this a little.

adamziel commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -98,26 +93,32 @@ export default function NavigationMenuSelector( {
-							} ) }
+							<MenuItemsChoice
+								value={ currentMenuId }
+								onSelect={ ( selectedId ) => {
```
nit: useCallback would be nice for that to avoid passing in a different object each time

adamziel reviewed (approved):
It works as intended. While I see some possible improvements, I don't think any of them are blocking and we could take them to another PR. Thank you for all your great work here @getdave !

scruffian reviewed (approved):
LGTM

getdave committed:
Remove isPlaceholder checks when already confirmed

getdave committed:
Optimise props

Addresses

- https://github.com/WordPress/gutenberg/pull/38907#discussion_r816784547
- https://github.com/WordPress/gutenberg/pull/38907#discussion_r816784643

getdave commented:
Last commit broke something 🤦

getdave committed:
Squash and rebase

getdave committed:
Check for possible empty menus values

dugyen commented:
The fix works perfectly on my end.
